python/ray/tests/test_reference_counting.py

```diff
@@ -0,0 +1,162 @@
+            wait_for_file(path)
+        # This should be spilled to plasma.
+        return large_object()
+
+    # Test that regular plasma dependency refcounts are decremented once the
+    # task finishes.
+    f = random_path()
+    large_dep = ray.put(large_object())
+    result = one_dep.remote(large_dep, path=f)
+    check_refcounts({large_dep: (1, 1), result: (1, 0)})
+    touch(f)
+    # Reference count should be removed once the task finishes.
+    check_refcounts({large_dep: (1, 0), result: (1, 0)})
+    del large_dep, result
+    check_refcounts({})
+
+    # Test that inlined dependency refcounts are decremented once they are
+    # inlined.
+    f = random_path()
+    dep = one_dep.remote(None, path=f)
+    check_refcounts({dep: (1, 0)})
+    result = one_dep.remote(dep)
+    check_refcounts({dep: (1, 1), result: (1, 0)})
+    touch(f)
+    # Reference count should be removed as soon as the dependency is inlined.
+    check_refcounts({dep: (1, 0), result: (1, 0)}, timeout=1)
+    del dep, result
+    check_refcounts({})
+
+    # Test that spilled plasma dependency refcounts are decremented once
+    # the task finishes.
+    f1, f2 = random_path(), random_path()
+    dep = one_dep_large.remote(None, path=f1)
+    check_refcounts({dep: (1, 0)})
+    result = one_dep.remote(dep, path=f2)
+    check_refcounts({dep: (1, 1), result: (1, 0)})
+    touch(f1)
+    ray.get(dep, timeout=5.0)
+    # Reference count should remain because the dependency is in plasma.
+    check_refcounts({dep: (1, 1), result: (1, 0)})
+    touch(f2)
+    # Reference count should be removed because the task finished.
+    check_refcounts({dep: (1, 0), result: (1, 0)})
+    del dep, result
+    check_refcounts({})
+
+    # Test that regular plasma dependency refcounts are decremented if a task
+    # fails.
+    f = random_path()
+    large_dep = ray.put(large_object())
+    result = one_dep.remote(large_dep, path=f, fail=True)
+    check_refcounts({large_dep: (1, 1), result: (1, 0)})
+    touch(f)
+    # Reference count should be removed once the task finishes.
+    check_refcounts({large_dep: (1, 0), result: (1, 0)})
+    del large_dep, result
+    check_refcounts({})
+
+    # Test that spilled plasma dependency refcounts are decremented if a task
+    # fails.
+    f1, f2 = random_path(), random_path()
+    dep = one_dep_large.remote(None, path=f1)
+    check_refcounts({dep: (1, 0)})
+    result = one_dep.remote(dep, path=f2, fail=True)
+    check_refcounts({dep: (1, 1), result: (1, 0)})
+    touch(f1)
+    ray.get(dep, timeout=5.0)
+    # Reference count should remain because the dependency is in plasma.
+    check_refcounts({dep: (1, 1), result: (1, 0)})
+    touch(f2)
+    # Reference count should be removed because the task finished.
+    check_refcounts({dep: (1, 0), result: (1, 0)})
+    del dep, result
+    check_refcounts({})
+
+
+if __name__ == "__main__":
+    import pytest
+    import sys
+    sys.exit(pytest.main(["-v", __file__]))
```

Comment:
Can we also add a "stress test" style test that fails without ref counting but succeed afterwards?